### PR TITLE
Add NetIQ ldap-flavour

### PR DESCRIPTION
**Pull request by m-bucher**
This is basically POSIX-LDAP with a view changes.

https://de.wikipedia.org/wiki/NetIQ_eDirectory

### CODE REVIEW

**Comment by lzap** [comment]
@ares @theforeman/core there are not enough active maintainers for this repo, it's a quiet one as you can tell from the history :-)

**Comment by m-bucher** [comment]
> Sure, this will require some packaging work, can you confirm you can create RPM/DEB packages for those dependencies? Mind adding PRs prior merging this?

@lzap not sure what you are referring to. The PR does not add new dependencies, so I would imagine this only requires a simple version bump of the already existing spec-files in foreman-packaging, or am I missing something?

**Review by lzap — APPROVED** [review]
Apologies, I seriously need to wear my glasses.

**Comment by lzap** [comment]
I need to ask someone from the core team to cut the release.

File: lib/ldap_fluff/netiq_member_service.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+  # AND or OR all of the filters together
+  def merge_filters(filters = [], all = false)
+    if !filters.nil? && filters.size >= 1
+      filter = filters[0]
+      filters[1..(filters.size - 1)].each do |gfilter|
+        filter = (all ? filter & gfilter : filter | gfilter)
+      end
+      return filter
+    end
```
Comment (adamruzicka) [inline lib/ldap_fluff/netiq_member_service.rb:72]: I found this a little bit hard to follow, but couldn't it be shortened to something like this?
```suggestion
    (filters || []).reduce(all ? :& : :|)
```

File: lib/ldap_fluff/netiq.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+class LdapFluff::NetIQ < LdapFluff::Generic
+
+  def bind?(uid = nil, password = nil, opts = {})
+    unless uid.include?(',') || opts[:search] == false
+      service_bind
+      user = @member_service.find_user(uid)
+      uid = user.first.dn if user && user.first
```
Comment (adamruzicka) [inline lib/ldap_fluff/netiq.rb:7]: Is the condition necessary? If I'm reading it right then https://github.com/theforeman/ldap_fluff/pull/71/files#diff-150a012639e8faaebe318dc258205828794c64cc2241d64e595326d501bd8959R13 either returns an array with at least one argument or raises an exception

File: lib/ldap_fluff/netiq_member_service.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    userlist.each do |current_user|
+       result << find_by_dn(current_user)[0][@attr_login][0]
+    end
+
+    result
+  end
+
+  # return an ldap user with groups attached
+  # note : this method is not particularly fast for large ldap systems
```
Comment (adamruzicka) [inline lib/ldap_fluff/netiq_member_service.rb:38]: Could we incorporate the changes from https://github.com/theforeman/ldap_fluff/pull/63 to make it faster?

File: lib/ldap_fluff/netiq_member_service.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    raise self.class::UIDNotFoundException if (user.nil? || user.empty?)
+    user
+  end
+
+  def get_logins(userlist)
+    userlist.map(&:downcase!)
+    result = []
+
+    userlist.each do |current_user|
+       result << find_by_dn(current_user)[0][@attr_login][0]
```
Comment (adamruzicka) [inline lib/ldap_fluff/netiq_member_service.rb:31]: Indentation looks off

**Review by adamruzicka — COMMENTED** [review]
I'm fairly new here so take this review more of a food for thought kind of thing.

In general it looks good, although I don't have an environment where I could test this.

File: lib/ldap_fluff/netiq_member_service.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  # note : this method is not particularly fast for large ldap systems
+  def find_user_groups(uid)
+    groups = []
+
+    @ldap.search(:filter => Net::LDAP::Filter.eq('memberuid', uid), :base => @group_base).each do |entry|
+      groups << entry[:cn][0]
+    end
+
+    user = find_user(uid)[0][:dn][0]
+    @ldap.search(:filter => Net::LDAP::Filter.eq('member', user), :base => @group_base).each do |entry|
```
Comment (ekohl) [inline lib/ldap_fluff/netiq_member_service.rb:47]: Can't you reduce this to a single search using an `or`? Like
```ruby
filter = Net::LDAP::Filter.eq('memberuid', uid) | Net::LDAP::Filter.eq('member', user)
```

File: lib/ldap_fluff/netiq_member_service.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    entry, base = search_dn.split(/(?<!\\),/, 2)
+    _entry_attr, entry_value = entry.split('=', 2)
+    entry_value = entry_value.gsub('\,', ',')
+    user = @ldap.search(:filter => name_filter(entry_value, "workforceid"), :base => base)
+    raise self.class::UIDNotFoundException if (user.nil? || user.empty?)
+    user
+  end
+
+  def get_logins(userlist)
+    userlist.map(&:downcase!)
```
Comment (ekohl) [inline lib/ldap_fluff/netiq_member_service.rb:27]: Is it a good idea to use an in place replacement? `downcase!` feels insecure. How about
```suggestion
    userlist = userlist.map(&:downcase)
```

Or combine it in the method below with `find_by_dn(current_user.downcase)`.

File: lib/ldap_fluff/netiq_member_service.rb (hunk 1 above)
Comment (m-bucher) [reply lib/ldap_fluff/netiq_member_service.rb:72]: The code is taken from `posix_member_service.rb` verbatim, so if this should be changed, we should do it for both.

**Comment by m-bucher** [comment]
DRYed the code removing the duplicates from `Posix`, and using it as base-class, which should 'fix' most of the issues related to copy-pasted code.
Also added (ok, copy-pasted from `Posix`) unit-tests for `NetIQ` for which I might want to do the same :innocent:

**Comment by m-bucher** [comment]
Since I did various changes I have yet to test them against the actual system next week :sweat_smile:

File: lib/ldap_fluff/netiq_member_service.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    user
+  end
+
+  def get_logins(userlist)
+    result = []
+    userlist.each do |current_user|
+      result << find_by_dn(current_user&.downcase)[0][@attr_login][0]
+    end
+
+    result
```
Comment (ekohl) [inline lib/ldap_fluff/netiq_member_service.rb:20]:
```suggestion
    userlist.map do |current_user|
      find_by_dn(current_user&.downcase)[0][@attr_login][0]
    end
```

File: lib/ldap_fluff/netiq_member_service.rb (hunk 4 above)
Comment (m-bucher) [reply lib/ldap_fluff/netiq_member_service.rb]: Now also checked by rubocop :sunglasses:

File: lib/ldap_fluff/netiq.rb (hunk 2 above)
Comment (m-bucher) [reply lib/ldap_fluff/netiq.rb]: Obsolete for this PR now, but still a good point for a new PR :wink:

**Review by ekohl — APPROVED** [review]
I don't have the setup to test this, but looks correct to me. Just a note on n+1 query complexity, though I'm not sure it needs to be blocking.

**Comment by m-bucher** [comment]
fixed rubocop finiding :innocent:

File: lib/ldap_fluff/netiq_member_service.rb (hunk 3 above)
Comment (ekohl) [reply lib/ldap_fluff/netiq_member_service.rb:27]: Given the updates, is this still slow? Or in other words, is the note still relevant or should it be removed?

Comment (m-bucher) [reply lib/ldap_fluff/netiq_member_service.rb:27]: I have not tested it, yet, but probably at least less slow as before :grin:

**Comment by sbernhard** [comment]
How do we want to proceed with this @ekohl / @lzap?

**Comment by lzap** [comment]
I have no problems with this, I do not actively work on Foreman so don't want to break something. I will let @ekohl to merge it. Thank you for the contribution! Apologies for the delay.

**Comment by ekohl** [comment]
Thanks!
